go/enclave/l2chain/l2_chain.go

```diff
@@ -543,6 +543,80 @@ func (oc *ObscuroChain) processState(batch *core.Batch, txs []*common.L2Tx, stat
 	return rootHash, executedTransactions, txReceipts, synthReceipts
 }
 
+// ResyncStateDB can be called to ensure stateDB data is available for the canonical L2 batch chain
+// After an (ungraceful) shutdown this method must be called to rebuild the stateDB data based on the persisted batches
+func (oc *ObscuroChain) ResyncStateDB() error {
+	batch, err := oc.storage.FetchHeadBatch()
+	if err != nil {
+		oc.logger.Info("no head batch found in DB after restart", log.ErrKey, err)
+		return nil
+	}
+	if !stateDBAvailableForBatch(oc.storage, batch.Hash()) {
+		oc.logger.Info("state not available for latest batch after restart - rebuilding stateDB cache from batches")
+		err = oc.replayBatchesToValidState()
+		if err != nil {
+			return fmt.Errorf("unable to replay batches to restore valid state - %w", err)
+		}
+	}
+	return nil
+}
+
+// replayBatchesToValidState is used to repopulate the stateDB cache with data from persisted batches. Two step process:
+// 1. step backwards from head batch until we find a batch that is already in stateDB catch, builds list of batches to replay
+// 2. iterate that list of batches from the earliest, process the transactions to calculate and cache the stateDB
+func (oc *ObscuroChain) replayBatchesToValidState() error {
+	// this slice will be a stack of batches to replay as we walk backwards in search of latest valid state
+	// todo: consider capping the size of this batch list using FIFO to avoid memory issues, and then repeating as necessary
+	batchesToReplay := make([]*core.Batch, 0)
```

Comment:
ooc, why make an empty slice ?

Reply:
Cool! I didn't know you could do that, didn't know append worked with nil slices.